man/dpath.8
.\" $Revision: 1.3 $
.TH DPATH 8
.SH NAME
dpath \- Diablo program to return commonly used paths for use in shell scripts
.PP
.SH SYNOPSIS
.B dpath [-C diablo.config] [-c] [-s] [-v] [-q] [id ...]

.SH DESCRIPTION
.IR DPath 
processes the diablo.config file and prints out all configurable paths
in a bourne-shell compatible format by default.   If you give it a single
path identifier, it returns the path for that identifier without anything
extra (i.e. suitable for backticked inclusion).
.PP
.B -c
\- Generate the output in csh style 'set' commands.
.PP
.B -s
\- Generate the output in bourne style 'var=data' commands.
.PP
.B -v
\- Generate verbose output (csh or sh style set variable).  This option
is implied by -c and -s.
.PP
.B -q
\- Prevent complaints from being printed to stderr.
.PP

.SH "SEE ALSO"
diablo(8), 
dicmd(8),
didump(8),
diload(8),
dnewslink(8),
doutq(8),
dexpire(8),
dexpireover(8),
diconvhist(8),
dilookup(8),
dspoolout(8),
dkp(8),
dpath(8),
diablo-kp(5),
diablo-files(5)
.PP



RELEASE_NOTES

			    RELEASE NOTES

		  DIABLO NEWS TRANSIT AND READER SYSTEM

		      diablo-users@openusenet.org
		    http://www.openusenet.org/diablo/
			   news.software.nntp

Known Issues:
	- OLD: dreaderd: problem with corrupt header files caused
	  by what seems to be a lock race condition with 'dexpireover -R'.
	  This problem has existed for years.

Experimental (features/utilities that are new and/or not properly tested
  	      and hence should be used with care on production systems):
	- Compressed spools
	- Compressed feeds
	- IPv6 support
	- dnewsfeeds %include option
	- dexpirecache
	- dcancel

Summary of changes likely to impact an upgrade from 5.x
	- new dreader header index format. Cannot go back to
	  older versions as they will not be able to read
	  the new version.

Summary of changes likely to impact an upgrade from 4.x
	- history version 2 (offsets converted to indexes)
	- new history expire (dhisexpire)
	- fixed bug in dexpire.ctl inheritance handling

Summary of changes likely to impact an upgrade from 3.x
	- New spool storage format. Each article is stored with a binary
	  header. This new format is not usable with older versions of
	  diablo, although 4.x can read prior storage formats.

Diablo 5-current

2009-05-25
	* dreaderd: Don't know why no one ever noticed that tcp
	  socket options per user weren't getting set right.

2009-04-20
        * dcancel: Add the ability to cancel articles by article
          number in the reader header index. (from Michael Dorrington)

2009-04-18
	* Oops, never committed the stunnel patch for useproxied.

2009-02-01
	* parsedate wouldn't deal well with certain days of the
	  week that weren't capitalized

2009-01-28
	* Cleanup of various code problems, including new expire
	  cleanup, spaces-in-Newsgroups cleanup, etc.

2009-01-15
	* Implement alternate article numbering views for clients
	  (this version still needs some cleanup).  Clients may
	  indicate their capabilities via the XNUMBERING extension.

	  RFC977 imposes no real limits on article numbering.  This
	  may cause some issues with non-64-bit-clean clients.

	  RFC3977 imposes a 2^31-1 limit AND places restrictions
	  on the resetting of the low-water mark, which is excessively
	  retarded because it effectively means that once a group hits
	  the limit, it's just dead, but it's the behaviour that is
	  defined in the RFC.

	  WINDOW guarantees compatibility with non-64-bit-clean
	  clients by restricting the article numbers to a window
	  between 1 and 2^31-1; further, the number of articles in
	  a group is limited to guarantee that most clients should
	  detect it as a "group reset" along the various lines of
	  ill-defined logic often used by clients.  It is not a
	  frequent event, but when it happens, this has the best
	  possibility of being successful without being too
	  disruptive.

	  Another patch shortly will clean this code up, and will
	  allow for selection of which view to present by default
	  to clients who do not explicitly request a particular
	  view.

2008-12-24
	* Absolutely huge changes that have been underway for more
	  than half a year.  

	  There are a set of interrelated but different problems
	  going on in the system.

	  The first is that the legacy overview code was not 64-bit
	  safe; this effectively limited overviews for a group to
	  around 60 million articles and then wackiness happened.  
	  The code has been walked through and updated so that on 
	  64-bit platforms, larger overviews are possible (we've 
	  exercised it to 500 mil). 32-bit platforms currently lose
	  due to the inability of the operating system to map large 
	  files, but support for this is also possible and a cleanly 
	  coded solution may be committed in the future.

	  Further, article numbering is a disaster waiting to happen.
	  The legacy code is organized to handle 31-bit message 
	  numbers.  Certain alt.binaries high volume groups are 
	  poised to pass message number 2^31-1.  Now, on one hand,
	  the problem could be solved by simply 64-bit-ifying the
	  code to handle 63-bit article numbers.  On the other hand,
	  some nimrods decided to specify 2^31-1 as a maximum article
	  number in RFC3977, which, fortunately, we don't really 
	  claim compliance with.  This code revision simply extends
	  the article numbering to 2^63.  In doing so, it will update
	  your overview files from VERSION 3 to VERSION 4.  This is
	  a harmless change except for groups where article numbering
	  exceeds 2^31-1.  Some effort was put into keeping the files
	  otherwise-compatible with older code.

	  Note that if you are using Diablo as an article numbering
	  master, you will need to update Diablo AND dreaderd.

	  People running this codebase will be able to have larger
	  overviews and won't run into a problem at artno 2^31-1.

2008-12-22
	* Various compiler warning cleanups

2008-12-19
	* Runaway spool thread status flag repurposing, based on a 
	  strategy POC'd by tim@euroaccess (thanks for the headaches, 
	  grin)

2008-12-18
	* Various compiler warning cleanups

2008-12-18
	* It's 2008.  Everyone has zlib.  If I don't hear objections, I
	  will probably be slowly removing conditional support for it.

2008-12-18
	* Add error checking for group resizing; many platforms limit
	  calloc memory availability and application of an "unlimit" or
	  equivalent in the daily.reader script might be required.  This
	  just adds a reasonable error message

2008-12-18
	* Add compressed header transmission to clients (XZVER support)
	  Newshosting and a few other places have been running this for
	  almost a year without any complaints, but it is still in
	  testing.

2008-12-16
	* drgpctl masscancel: support cancellation of large numbers of
	  articles from very large groups (one-pass through the overview
	  file)

2008-12-15
	* Dump threads status from sigUsr1 (can't find who submitted this
	  patch, apologies, whoever)

2008-12-15
	* Report virtual server name in log lines


util/dpath.c

/*
 * DPATH.C	- Return the path to specific diablo files.  Typically used
 *		  by shell scripts.
 *
 * dpath [-C config] [-v] [-q] [-c] idname
 *
 *	(see top of diablo.config for id names)
 */

#include "defs.h"

void PrintDiabloPath(const char *ptr);

int VerboseOpt = 0;
int QuietOpt = 0;
int CshStyleOpt = 0;

int
main(int ac, char **av)
{
    int i;
    int loop;
    int pcount = 0;

    LoadDiabloConfig(ac, av);

    for (loop = 0; loop < 2; ++loop) {
	for (i = 1; i < ac; ++i) {
	    char *ptr = av[i];

	    if (*ptr != '-') {
		if (loop == 1) {
		    PrintDiabloPath(ptr);
		    ++pcount;
		}
		continue;
	    }
	    ptr += 2;
	    switch(ptr[-1]) {
	    case 'C':
		/*
		 * -C configfile	(handled by LoadDiabloConfig)
		 */
		if (*ptr == 0)
		    ++i;
		break;
	    case 'c':
		CshStyleOpt = 1;
		VerboseOpt = 1;
		break;
	    case 'q':
		QuietOpt = 1;
		break;
	    case 's':
		CshStyleOpt = 0;
		VerboseOpt = 1;
		break;
	    case 'V':
		PrintVersion();
		break;
	    case 'v':
		VerboseOpt = 1;
		break;
	    default:
		fprintf(stderr, "Bad option: %s\n", ptr - 2);
		exit(1);
	    }
	}
    }
    if (pcount == 0) 
	PrintDiabloPath(NULL);
    return(0);
}

typedef const char *ccptr;

typedef struct PathList {
    const char *id;
    const char **pglob;
    const char *(*func)(const char *pat);
} PathList;

PathList PList[] = {
    { "home",	&NewsHome, NULL },
    { "spool",	&SpoolHomePat, PatExpand },
    { "dqueue",	&DQueueHomePat, PatExpand },
    { "group",	&GroupHomePat, PatExpand },
    { "cache",	&CacheHomePat, PatExpand },
    { "feeds",	&FeedsHomePat, PatExpand },
    { "log",	&LogHomePat, PatExpand },
    { "lib",	&LibHomePat, PatExpand },
    { "db",	&DbHomePat, PatExpand },
    { "run",	&RunHomePat, PatExpand },
    { "diablo_socket",	&DiabloSocketPat, PatRunExpand },
    { "dreader_socket",	&DReaderSocketPat, PatRunExpand },
    { "feednotify",	&DFeedNotifySocketPat, PatRunExpand },


    { "dexpire",	&DExpireCtlPat, PatLibExpand },
    { "dcontrol",	&DControlCtlPat, PatLibExpand },
    { "diablo_hosts",	&DiabloHostsPat, PatLibExpand },
    { "dserver_hosts",	&DServerHostsPat, PatLibExpand },
    { "moderators",	&ModeratorsPat, PatLibExpand },
    { "dnewsfeeds",	&DNewsfeedsPat, PatLibExpand },
    { "dnntpspool",	&DNNTPSpoolCtlPat, PatLibExpand },
    { "distrib_pats",	&DistribDotPatsPat, PatLibExpand },
    { "distributions",	&DistributionsPat, PatLibExpand },

    { "server_dactive",	&ServerDActivePat, PatDbExpand },
    { "reader_dactive",	&ReaderDActivePat, PatDbExpand },
    { "dhistory",	&DHistoryPat, PatDbExpand },
    { "spam.body.cache",	&SpamBodyCachePat, PatDbExpand },
    { "spam.nph.cache",	&SpamNphCachePat, PatDbExpand },
    { "pcommit_cache",	&PCommitCachePat, PatDbExpand }
};

void
PrintDiabloPath(const char *ptr) 
{
    int i;
    int c = 0;

    for (i = 0; i < arysize(PList); ++i) {
	PathList *pl = &PList[i];

	if (ptr == NULL || strcmp(ptr, pl->id) == 0) {
	    if (VerboseOpt) {
		switch(CshStyleOpt) {
		case 0:
		    printf("%s=", pl->id);
		    break;
		case 1:
		    printf("set %s = \"", pl->id);
		    break;
		}
	    }

	    if (pl->func)
		printf("%s", pl->func(*pl->pglob));
	    else
		printf("%s", *pl->pglob);

	    if (CshStyleOpt == 1)
		printf("\"");
	    printf("\n");
	    ++c;
	}
    }
    if (c == 0) {
	if (QuietOpt == 0)
	    fprintf(stderr, "dpath: illegal path id '%s'\n", ptr);
    }
}

